Ahoyapps 933 add passcode

timmydoza opened the pull request:
<!-- Describe your Pull Request. You may remove some parts that are not applicable. -->

**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

## Pull Request Details

### JIRA link(s):

- [AHOYAPPS-933](https://issues.corp.twilio.com/browse/AHOYAPPS-933)

### Description

This PR adds extra random digits to the deployed serverless URL. It also adds some logic so that the app will expire one week after being deployed. 

This is what is printed to the console after a deployment:
```
App deployed to: https://rtc-diagnostics-mk8gr9b4-5753-dev.twil.io
This URL is for demo purposes only. It will expire on Mon Nov 30 2020 12:52:32 GMT-0700
```

Image of UI of an expired app:
![image](https://user-images.githubusercontent.com/12685223/100008391-a2808300-2d8a-11eb-944d-093e3a7c1ada.png)

Image of tooltip content:
![image](https://user-images.githubusercontent.com/12685223/100008420-af9d7200-2d8a-11eb-874f-49980e416808.png)

These warnings are displayed to the user after they try to run the network tests.

Edit - I updated the Icon to better match the designs:
![image](https://user-images.githubusercontent.com/12685223/100132882-475e9700-2e43-11eb-9a80-b334e88672dd.png)




## Burndown

### Before review
* [ ] Updated CHANGELOG.md if necessary
* [x] Added unit tests if necessary
* [ ] Updated affected documentation
* [ ] Verified locally with `npm test`
* [x] Manually sanity tested running locally
* [x] Included screenshot as PR comment (if needed)
* [x] Ready for review

### Before merge
* [ ] Got one or more +1s
* [ ] Re-tested if necessary

charliesantos commented on serverless/assets/verify_expiry.private.js:
```diff
@@ -0,0 +1,15 @@
+    });
+    return callback(null, response);
+  }
+};
```
Curious why is this under assets and the rest are under functions?

charliesantos replied:
If you agree with my middleware suggestion, we can probably move this under `middleware` folder.

timmydoza replied:
Uploading JS files as assets allows us to `require()` them in our functions. This is what is recommended in [the docs](https://www.twilio.com/docs/labs/serverless-toolkit/general-usage#assets):

>Assets are a great way for you to also host resources that you want to access from your Functions that are not other Functions or should not be available to the public. For example configuration files or other JavaScript files you want to require(). You can store those as private Assets.

The same technique is used in the plugin-rtc repo, and both you and Aaron have suggested that the folder be named `/middleware`: https://github.com/twilio-labs/plugin-rtc/pull/55#discussion_r528928922

I think this makes sense. The file is technically an asset, but it functions as middleware, so I'll change the directory name.

charliesantos commented on serverless/assets/verify_expiry.private.js:
```diff
@@ -0,0 +1,15 @@
+exports.handler = function (context, event, callback) {
```
Nice, I see what you're doing here. It seems like this is acting like a middleware to your endpoints. However, the implementation requires duplicating the code block below to all your endpoints, which is not ideal.

```js
  const verifyExpiry = require(Runtime.getAssets()['/verify_expiry.js'].path);
  verifyExpiry.handler(context, event, callback);
```

How about we do this? Update this file to act like a middleware:
```js
const authMiddleware = function(handler) {
  return function(context, event, callback) {
    if (Date.now() > context.APP_EXPIRY) {
      const response = new Twilio.Response();
      response.appendHeader('Content-Type', 'application/json');
      response.setStatusCode(401);
      response.setBody({
        error: {
          message: 'passcode expired',
          explanation:
            'foo bar.',
        },
      });
      return callback(null, response);
    }
    handler(context, event, callback);
  };
};
```

Then in your endpoints handlers, for example `serverless/functions/app/turn-credentials.js`, you can use it like so:

```js
exports.handler = authMiddleware(function(context, event, callback) {
  const client = context.getTwilioClient();
  client.tokens.create({ ttl: 30 }).then((token) => callback(null, token));
});
```

timmydoza replied:
I see what you're saying.  We would still need to require() the `authMiddleware` function in each file with:

`const authMiddleware = require(Runtime.getAssets()['/auth_middleware.js'].path);`

So this will only save us one line of code per function. On one hand, I don't think it is worth doing for only two functions (saving two lines of code), but on the other hand, it is a little cleaner. I'll think about this one...

charliesantos replied:
Ok no worries. I don't have a strong opinion on this because we only have two endpoints at the moment. But it's good to have something cleaner if we end up building more on top of this.

timmydoza replied:
I think I'll keep it as is, but this `authMiddleware` function is a good pattern to keep in mind.

timmydoza commented on package.json:
```diff
@@ -14,6 +14,8 @@
     "dotenv": "^8.2.0",
     "lodash": "^4.17.20",
     "loglevel": "^1.6.8",
+    "moment": "^2.29.1",
```
Yeah, now that I think about this one a little more I think we can just use vanilla JS.

charliesantos commented on serverless/functions/app/token.js:
```diff
@@ -1,4 +1,7 @@
+  const verifyExpiry = require(Runtime.getAssets()['/verify_expiry.js'].path);
+  verifyExpiry.handler(context, event, callback);
+
   const AccessToken = Twilio.jwt.AccessToken;
```
I just realized, if we keep this as is (which is ok to me), how do we make sure the rest of the handler code doesn't execute if `verifyExpiry.handler` returns a failure code? In the current implementation, lines 5 onwards will still execute even if `verifyExpiry.handler` executes the callback.

timmydoza replied:
Great question. I think we are ok. Per the [docs](https://www.twilio.com/docs/runtime/functions/invocation):

> When your Twilio Function code has completed, you call the callback() method to emit your response. After executing the callback() method, your Twilio Function execution will be terminated. This includes any asynchronous processes that may still executing.

charliesantos replied:
Ok cool.

timmydoza commented:
@charliesantos This is ready for another review. I have:

- Merged in main and resolved conflicts
- Removed moment
- Changed `assets/` to `middleware/`

timmydoza commented on src/NetworkTestWidget/EdgeResult/EdgeResult.test.tsx:
```diff
@@ -57,19 +57,24 @@ describe('the EdgeResult component', () => {
-    codecPreferences: [PCMU, Opus],
-    label: 'Ashburn (PCMU, Opus)',
-  }].forEach(test => {
+  [
```
@charliesantos Not a big deal, but I'm not sure if `prettier` is working right for you. These files should be automatically be formatted on each commit by the husky git hooks: https://github.com/twilio/rtc-diagnostics-react-app/blob/main/package.json#L86

Not sure why this would be. This should all be set up just by running `npm install`.

charliesantos reviewed (approved):
🚢  🇮🇹
